Dolci/schedules application

Ig-dolci opened the pull request.

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,899 @@
+    "J = sum(u_2.reshape(-1) * 0.5) * 1/burger.model[\"nx\"]  # Compute the functional value.\n",
+    "E, h = taylor_remainder_test(burger.adjoint_work_memory, J, u0)  # Compute the Taylor remainder.\n",
+    "convergence_rates(E, h, show=True)\n",
+    "senstivity = sum(burger.adjoint_work_memory[StorageType.WORK][0].reshape(-1) * u0.reshape(-1)) * 1 / model[\"nx\"]\n",
```
```suggestion
    "senstivity = sum(burger.adjoint_work_memory[StorageType.WORK][0].reshape(-1) * u0.reshape(-1)) / model[\"nx\"]\n",
```
Is the `1 / model["nx"]` factor needed?

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,899 @@
+    "\n",
+    "The purpose of this adjoint-based sensitivity computation is to use every checkpointing approach available in the `checkpoint_schedules` package and verify the consistent of the results. We employ a fundamental tool used in verification of gradients, which is the\n",
+    "*second order Taylor remainder convergence test*. Thus, let us consider the functional $I(u_0)$  and $\\nabla_{u_0}$ be its gradient with respect to the control parameter $u_0$. Let $u$ be the solution of the forward problem, and let $\\delta u$ be a perturbation to $u$. Therefore, the *Taylor remainder convergence test* is given by the expression: \n",
+    "$$ \\left|I(u + \\delta u) - I(u) - \\nabla_{u_0} I \\cdot \\delta u\\right| \\rightarrow 0 \\quad \\mathrm{at} \\ O(\\delta u^2).$$\n",
```
The perturbation should be $\varepsilon \delta u$ with scalar $\varepsilon$ and direction $\delta u$, and then the Taylor remainder is $O(\varepsilon^2)$.

jrmaddison reviewed (changes requested):
Some general comments:

- I think there is an error in the forward residual -- the boundary conditions don't seem to be applied.
- The adjoint derivation is motivated by a continuous adjoint approach, but the later Taylor testing verifies discrete adjoint consistency.
- The displayed adjoint sensitivity is a directional derivative rather than the full gradient.
- I recommend reducing the number of example schedules.
- I recommend removing all the code comments in the example schedules.
- The disk storage schedules don't actually store checkpoints on disk.
- I don't understand the `single_storage` behaviour.

Ig-dolci commented:
> Some general comments:
> 
> * I think there is an error in the forward residual -- the boundary conditions don't seem to be applied.
> * The adjoint derivation is motivated by a continuous adjoint approach, but the later Taylor testing verifies discrete adjoint consistency.
> * The displayed adjoint sensitivity is a directional derivative rather than the full gradient.
> * I recommend reducing the number of example schedules.
> * I recommend removing all the code comments in the example schedules.
> * The disk storage schedules don't actually store checkpoints on disk.
> * I don't understand the `single_storage` behaviour.

1. The boundary conditions were fixed.
2. I moved the code and the text for discrete adjoint.
3. I reduced the number of example schedules.
4. The disk storage is now used.
5. `sinlge_storage` storage is not used anymore.
6. I removed the docstring of the private methods.

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,755 @@
+    "        The spatial mesh.\n",
+    "    \"\"\"\n",
+    "    def __init__(self, model, forward_initial_condition, mesh):\n",
+    "        self.model = model\n",
```
```suggestion
    "        self.model = dict(model)\n",
```

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,755 @@
+    "        M = self._mass_matrix()\n",
+    "        K = self._stiffness_matrix()\n",
+    "        # Get the initial condition\n",
+    "        u = self.forward_work_memory[StorageType.WORK][n0]\n",
+    "        del self.forward_work_memory[StorageType.WORK][n0]\n",
```
```suggestion
    "        u = self.forward_work_memory[StorageType.WORK].pop(n0)\n",
```

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,755 @@
+    "        if write_ics:\n",
+    "            self._store_data(u, n0, storage, write_adj_deps, write_ics)\n",
+    "        u_new = copy.deepcopy(u)\n",
+    "        n1 = min(n1, self.model[\"max_n\"])\n",
+    "        step = n0\n",
+    "        while step < n1:\n",
```
`for step in range(n0, min(n1, self.model["max_n"]):`

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,755 @@
+    "            def _residual(u_new):\n",
+    "                C = self._advection_matrix(u_new)\n",
+    "                residual = (M + self.model[\"dt\"] * (- K + C)).dot(u_new) - M.dot(u)\n",
+    "                residual[0] = residual[self.model[\"nx\"] - 1] = 0\n",
```
```
residual[0] = u_new[0]
residual[self.model["nx"] - 1] = u_new[self.model["nx"] - 1]
```

Ig-dolci replied:
I kept `residual[0] = residual[self.model[\"nx\"] - 1] = 0` because the I lose the correct convergence rate with the modification asked above.

jrmaddison replied:
This is needed for an initial guess inconsistent with the boundary conditions. With other edits in the latest review I see second order convergence.

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,755 @@
+    "        if n1 <= self.model[\"max_n\"]:\n",
+    "            if n1 == self.model[\"max_n\"]:\n",
+    "                self._initialize_adjoint()\n",
+    "            u = self.adjoint_work_memory[StorageType.WORK][n1]\n",
+    "            del self.adjoint_work_memory[StorageType.WORK][n1]\n",
```
```suggestion
    "            u = self.adjoint_work_memory[StorageType.WORK].pop(n1)\n",
```

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,755 @@
+    "                self._initialize_adjoint()\n",
+    "            u = self.adjoint_work_memory[StorageType.WORK][n1]\n",
+    "            del self.adjoint_work_memory[StorageType.WORK][n1]\n",
+    "        step = n1\n",
+    "        M = self._mass_matrix()\n",
+    "        while step > n0:\n",
```
```suggestion
    "        M = self._mass_matrix()\n",
    "        for step in range(n1 - 1, n0  - 1, -1)\n"
```

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,755 @@
+    "                return residual\n",
+    "            u_new = self._solve_newton(u, _residual)\n",
+    "            u = copy.deepcopy(u_new)\n",
+    "            step += 1\n",
```
Delete

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,755 @@
+    "            u[0] = u[self.model[\"nx\"] - 1] = 0\n",
+    "            if clear_adj_deps:\n",
+    "                del self.adjoint_dependency[StorageType.WORK][step - 1]\n",
+    "            step -= 1\n",
```
Delete

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,755 @@
+    "\n",
+    "    def _initialize_adjoint(self):\n",
+    "        J = self._jacobian(self.forward_final_solution, self.model[\"max_n\"], self.model[\"max_n\"])\n",
+    "        M = self._mass_matrix()\n",
```
Delete

jrmaddison reviewed (commented):
Reviewed up the start of 'Checkpointing'. I haven't yet checked the numerics or that the schedules are being used as expected.

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,755 @@
+    "    \n",
+    "    def _solve_newton(self, u_prev, residual, tol=1e-8, max_iter=50):\n",
+    "        i = 0\n",
+    "        u_prev[0] = u_prev[-1] = 0\n",
+    "        u = copy.deepcopy(u_prev)\n",
```
```suggestion
    "        u = copy.deepcopy(u_prev)\n",
    "        u[0] = u[-1] = 0\n",
```

Ig-dolci replied:
I lose the correct convergence rate with the modification asked above.

jrmaddison replied:
See previous comment re. the Newton solver residual.

jrmaddison reviewed (commented):
From checking mass and stiffness matrices:

- There is an off-by-on error in the definition of $h$, which should be `model["lx"] / (model["nx"] - 1)`. This appears in three places.
- The `_mass_matrix` method returns the mass divided by $h$, which could be noted in a comment.
- The `_stiffness_matrix` method returns the stiffness matrix multiplied by `-\model["nu"] / h`. There are sign errors when the result is used  in `_residual` and `_jacobian`.

Ig-dolci commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,783 @@
+    "The time sequence of the adjoint system for the functional (Eq. (1)) is given by:\n",
+    "$$\n",
+    "\\begin{align*}\n",
+    "    \\lambda^{N+1} &= \\frac{\\partial I}{\\partial u^{N + 1}} \\\\\n",
```
I am still thinking about how I should write this sequence of adjoint equations.

Ig-dolci commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,755 @@
+    "$$\\frac{\\mathrm{d}\\widehat{I}}{\\mathrm{d}u_0} = \\frac{\\partial I}{\\partial u} \\frac{\\mathrm{d}u}{\\mathrm{d}u_0}. \\tag{3}$$\n",
+    "Computing $\\partial J/\\partial u$ is straightforward since the functional (Eq. (1)) is written in terms of the solution variable $u$. In contrast, the Jacobian $\\mathrm{d}u/\\mathrm{d}u_0$ involves an expensive computational operation that is avoided by using the adjoint method.\n",
+    "\n",
+    "In summary, the adjoint system is defined by using the relation of the functional $I$ with the forward equation $F(u, u_0)$. Taking the total derivative of $F(u, u_0)$ with respect to $u_0$ yields a relationship for the solution Jacobian $\\mathrm{d}u/\\mathrm{d}u_0$:\n",
```
I added references below at the line 72.

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,788 @@
+    }
+   ],
+   "source": [
+    "burger = BurgersEquation(model, u0, mesh)  # Create the burger's equation object.\n",
+    "burger.forward(0, model[\"max_n\"], storage=StorageType.WORK, write_adj_deps=True)  # Solve the forward problem.\n",
+    "burger.adjoint(0, model[\"max_n\"], clear_adj_deps=True)  # Solve the adjoint problem.\n",
+    "base_gradient = burger.gradient()  # Compute the gradient.\n",
+    "h = np.random.rand(model[\"nx\"])  # Random perturbation.\n",
+    "functional = burger.functional()\n",
+    "E, epsilon = taylor_remainder_test(burger.gradient(), functional, u0, h)  # Taylor remainder test.\n",
+    "convergence_rates(E, epsilon)  # Compute the convergence rates."
```
I don't think the comments are needed here (and similarly later)

Ig-dolci replied:
I removed the comments.

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,788 @@
+    "burger.forward(0, model[\"max_n\"], storage=StorageType.WORK, write_adj_deps=True)  # Solve the forward problem.\n",
+    "burger.adjoint(0, model[\"max_n\"], clear_adj_deps=True)  # Solve the adjoint problem.\n",
+    "base_gradient = burger.gradient()  # Compute the gradient.\n",
+    "h = np.random.rand(model[\"nx\"])  # Random perturbation.\n",
```
Should be deterministic, so that the example gives the same results when rerun.

Ig-dolci replied:
I moved to `delta_u = np.ones(model["nx"])`.

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,788 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We now compute the adjoint-based gradient using `SingleMemoryStorageSchedule` checkpointing approach. The `SingleMemoryStorageSchedule` stores the forward data of each time-step in working memory. As explained in the [notebook with illustrative example](https://nbviewer.org/github/firedrakeproject/checkpoint_schedules/blob/main/docs/notebooks/tutorial.ipynb), this schedule does not require t ...
```
This is conceptually the same as the previous case -- except that a scheduler is used to implement it.

Ig-dolci replied:
I removed this code cell.

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,788 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The following examples use `Revolve` [3] and `HRevolve` schedule [4]. The `Revolve` algorithm requires the definition of the maximal step `model[\"max_n\"]` before the execution of the forward solver, and also the specification of the number of checkpoints stored in memory (`model[\"max_n\"]`). Whereas `Revolve` allows only the storage in memory, the `Hrevolve` algorithm allows both disk and ...
```
```suggestion
    "The following examples use `Revolve` [3] and `HRevolve` schedules [4]. The `Revolve` algorithm requires the definition of the maximal step `model[\"max_n\"]` before the execution of the forward solver, and also the specification of the number of checkpoints stored in memory (`model[\"max_n\"]`). Whereas `Revolve` allows only the storage in memory, the `HRevolve` algorithm allows both disk and memory storage. The values `model[\"max_n\"]`, `model[\"chk_ram\"]` and `model[\"chk_dins\"]` are required to define the `HRevolve` schedule."
```

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,788 @@
+    "        M[0, 0] = M[-1, -1] = 1\n",
+    "        for step in range(n1, n0, - 1):\n",
+    "            J_T = self._jacobian(self.adjoint_dependency[StorageType.WORK][step]).T\n",
+    "            u_adj_coef = spsolve(J_T, u_adj)\n",
+    "            u_adj_new = M.T.dot(u_adj_coef)\n",
```
```suggestion
    "            u_adj_coef = spsolve(J_T, u_adj)\n",
    "            u_adj_coef[0] = u_adj_coeff[-1] = 0\n",
    "            u_adj_new = M.T.dot(u_adj_coef)\n",
```
Adjoint boundary conditions, assuming we don't have a boundary control.

Ig-dolci replied:
Done.

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,788 @@
+    "            The adjoint-based gradient.\n",
+    "        \"\"\"\n",
+    "        grad = self.adjoint_work_memory[StorageType.WORK][0]\n",
+    "        grad[0] = grad[self.model[\"nx\"] - 1] = 0\n",
```
Delete. Needed for an initial condition (or perturbation direction) which is non-zero on the boundary.

Ig-dolci replied:
This boundary conditions was removed.

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,788 @@
+    "            i += 1\n",
+    "            if np.linalg.norm(delta_u) < tol:\n",
+    "                break\n",
+    "        if i == max_iter:\n",
+    "            print(\"Newton's method did not converge in {} iterations.\".format(max_iter))\n",
```
```suggestion
    "            if i == max_iter:\n",
    "                print(\"Newton's method did not converge in {} iterations.\".format(max_iter))\n",
```

jrmaddison reviewed (changes requested):
Code should also be linted (e.g. using ruff).

Ig-dolci commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,788 @@
+    "        step += 1\n",
+    "        if step == self.model[\"max_n\"]:\n",
+    "            self.forward_final_solution = copy.deepcopy(u_new)\n",
+    "            self.adjoint_dependency[StorageType.WORK][step] = copy.deepcopy(u_new)\n",
```
done.

Ig-dolci commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,788 @@
+    "                    del self.adjoint_dependency[from_storage][step]\n",
+    "                if step in self.restart_forward[from_storage]:\n",
+    "                    del self.restart_forward[from_storage][step]\n",
+    "\n",
```
Done.

Ig-dolci commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,788 @@
+    "        num_nodes = self.model[\"nx\"]\n",
+    "        # u0 and u1 are the coefficients of the basis functions at the nodes.\n",
+    "        u0, u1 = symbols(\"u0 u1\") \n",
+    "        coeficients = Matrix([u0, u1]) \n",
```
Done.

Ig-dolci commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,788 @@
+    "                K_local[i, j] = b * integrate(dphi_deta[i] * dphi_deta[j], (\"eta\", 0, 1))\n",
+    "        return self._assemble_global_matrix(K_local)\n",
+    "    \n",
+    "    def _advection_matrix(self, u):\n",
```
Done.

Ig-dolci commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,788 @@
+    "        # u0 and u1 are the coefficients of the basis functions at the nodes.\n",
+    "        u0, u1 = symbols(\"u0 u1\")\n",
+    "        phi = self._basis_function()\n",
+    "        coeficients = Matrix([u0, u1]) \n",
```
Done.

Ig-dolci commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,788 @@
+    "            if write_adj_deps:\n",
+    "                self.adjoint_dependency[storage][step] = copy.deepcopy(data)\n",
+    "            if write_ics:\n",
+    "                self.restart_forward[storage][step] = copy.deepcopy(data)\n",
```
Done.

Ig-dolci commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,788 @@
+    "\n",
+    "    def _store_on_disk(self, data, step, adj_deps):\n",
+    "        if adj_deps:\n",
+    "            file_name = \"adj_dependency/fwd_\"+ str(step) +\".npy\"\n",
```
I added `tempfile.gettempdir`.

Ig-dolci commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,788 @@
+    "            with open(file_name, \"wb\") as f:\n",
+    "                np.save(f, data)\n",
+    "        else:\n",
+    "            file_name = \"restart_forward/fwd_\"+ str(step) +\".npy\"\n",
```
I added tempfile.gettempdir.

Ig-dolci commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,788 @@
+    "$$\\frac{\\mathrm{d}\\widehat{I}}{\\mathrm{d}u_0} = \\frac{\\partial I}{\\partial u} \\frac{\\mathrm{d}u}{\\mathrm{d}u_0}. \\tag{3}$$\n",
+    "Computing $\\partial J/\\partial u$ is straightforward since the functional (Eq. (1)) is written in terms of the solution variable $u$. In contrast, the Jacobian $\\mathrm{d}u/\\mathrm{d}u_0$ involves an expensive computational operation that is avoided by using the adjoint method.\n",
+    "\n",
+    "In summary, the adjoint system is defined by using the relation of the functional $I$ with the forward equation $F(u, u_0)$. Taking the total derivative of $F(u, u_0)$ with respect to $u_0$ yields a relationship for the solution Jacobian $\\mathrm{d}u/\\mathrm{d}u_0$:\n",
```
This text was removed.

Ig-dolci commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,788 @@
+    "where $\\lambda^{n}$ is the adjoint variable at time step $n$.\n",
+    "\n",
+    "The adjoint system is solved backward in time, we can verify this by taking the adjoint of Eq. (5), where at the time step zero we have two unknowns, $\\lambda^{0}$ and $\\lambda^{1}$, and at the time step $N$ we have only one unknown, $\\lambda^{N+1}$.\n",
+    "\n",
```
Removed.

Ig-dolci commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,788 @@
+   "metadata": {},
+   "source": [
+    "## Burger's equation solvers\n",
+    "We define a `BurgersEquation` class to compute the `forward` and `adjoint` solvers. In addition, the `BurgersEquation` class defines a `copy_data` method that copies the data from one storage type to another, and an `adjoint_initial_condition` method that initializes the adjoint.\n",
```
Done.

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,745 @@
+    "$$\\frac{\\mathrm{d}\\widehat{I}}{\\mathrm{d}u_0} = \\frac{\\partial I}{\\partial u} \\frac{\\mathrm{d}u}{\\mathrm{d}u_0}. \\tag{3}$$\n",
+    "Computing $\\partial J/\\partial u$ is straightforward since the functional (Eq. (1)) is written in terms of the solution variable $u$. In contrast, the Jacobian $\\mathrm{d}u/\\mathrm{d}u_0$ involves an expensive computational operation that is avoided by using the adjoint method.\n",
+    "\n",
+    "Taking the adjoint (Hermitian transpose) of the above equation:\n",
```
Missing an equation? Suggest citing  https://doi.org/10.1137/1.9780898718720 section 2.3.3, which looks like the same derivation.

Ig-dolci replied:
By mistake, I removed texts on the latest commits. This is now fixed. The reference  https://doi.org/10.1137/1.9780898718720  was added. I verified the book section 2.3.3.

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,745 @@
+    "which is the adjoint system. Finally, we obtain the following expression for the adjoint-based gradient computation:\n",
+    "$$\\frac{\\mathrm{d}\\widehat{I}}{\\mathrm{d} u_0}^* = - \\frac{\\partial F}{\\partial u_0}^* \\lambda.$$\n",
+    "In this current case, Burger's equation is written in weak form and discretized in time with the backward finite difference method, we have:\n",
+    "$$ F(u^{n+1}, u^n, v) = \\int_{\\Omega} u^{n + 1} \\cdot v - u^{n}  \\cdot v + \\Delta t \\left(u^{n + 1} \\frac{\\partial u^{n + 1}}{\\partial x} \\cdot v + \\frac{\\partial u^{n + 1}}{\\partial x}  \\cdot \\frac{\\partial v}{\\partial x}\\right) \\, dx = 0 \\quad \\forall v\\in V \\tag{4},$$\n",
```
Missing $\nu$.

Ig-dolci replied:
done.

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,745 @@
+    "$$\n",
+    "where $\\lambda^{n}$ is the adjoint variable at time step $n$.\n",
+    "\n",
+    "For additional details on the discrete adjoint-based gradient computation, you may refer to the following references [1, 2]."
```
[1] not needed if Gunzburger reference is added above.

Ig-dolci replied:
I removed this text and added Gunzburger reference above.

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,745 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Burger's equation solvers\n",
```
```suggestion
    "## Burger's equation solver\n",
```

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,745 @@
+    "## Burger's equation solvers\n",
+    "We define a `BurgersEquation` class to compute the `forward` and `adjoint` solutions, and to handle storing of forward and adjoint data.\n",
+    "\n",
+    "The Burger's equation is discretised using the Finite Element Method (FEM). We use continuous first-order Lagrange basis functions to discretise the spatial domain. `BurgersEquation` uses the Sympy Python library to assemble the matrices from the weak form of the Burger's equation."
```
```suggestion
    "The Burger's equation is discretised using the Finite Element Method (FEM). We use continuous first-order Lagrange basis functions to discretise the spatial domain. `BurgersEquation` uses the SymPy Python library to assemble the matrices from the weak form of the Burger's equation."
```

jrmaddison commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,745 @@
+    "    addition, it contains methods to copy data from one storage to another, and\n",
+    "    to set the initial condition for the adjoint.\n",
+    "    \"\"\"\n",
+    "    def __init__(self, schedule, equation):\n",
```
Maybe `solver` instead of `equation`.

Ig-dolci commented on docs/notebooks/burger.ipynb:
```diff
@@ -0,0 +1,745 @@
+    "            If `True`, the adjoint dependency data will be cleared.\n",
+    "        \"\"\"\n",
+    "        self.mode = \"adjoint\"\n",
+    "        if n1 <= self.model[\"max_n\"]:\n",
```
This was a mistake. I removed it.
